pkg/rng/entropy.go

```diff
@@ -0,0 +1,160 @@
+	// HW random number generator
+	HwRandomCurrentFile = "/sys/class/misc/hw_random/rng_current"
+	// HwRandomAvailableFile shows the current available
+	// HW random number generator
+	HwRandomAvailableFile = "/sys/class/misc/hw_random/rng_available"
+	// RandomEntropyAvailableFile shows how much of the entropy poolsize is used
+	RandomEntropyAvailableFile = "/proc/sys/kernel/random/entropy_avail"
+	// EntropyFeedTime sets the loop time for seeding /dev/random by /dev/hwrng
+	// in seconds
+	EntropyFeedTime time.Duration = 2
+	// EntropyBlockSize sets the bytes to read per Read function call
+	EntropyBlockSize = 128
+	// EntropyThreshold is used to stop seeding at specific entropy level
+	EntropyThreshold uint64 = 3000
+	// RandomDevice is the linux random device
+	RandomDevice = "/dev/random"
+	// HwRandomDevice is the linux hw random device
+	HwRandomDevice = "/dev/hwrng"
+)
+
+// Can be extended but keep in mind to priorize
+// more secure random sources like hw random over
+// timer, jitter based mechanisms. At the top of the array
+// is the highest priority.
+// <rng-name>
+var trngList = []string{
+	"tpm-rng",
+	"intel-rng",
+	"amd-rng",
+	"timeriomem-rng",
+}
+
+// Searches for available True Random Number Generator
+// inside the kernel api and sets the most secure on if
+// available which seeds /dev/hwrng
+func setAvailableTRNG() (bool, error) {
+	var (
+		currentRNG    string
+		availableRNGs []string
+		selectedRNG   string
+	)
+
+	availableFileData, err := ioutil.ReadFile(HwRandomAvailableFile)
+	if err != nil {
+		return false, err
+	}
+	availableRNGs = strings.Split(string(availableFileData), " ")
+
+	for _, trng := range trngList {
+		for _, rng := range availableRNGs {
+			if trng == rng {
+				selectedRNG = trng
+				break
+			}
+		}
+	}
+
+	if selectedRNG == "" {
+		return false, errors.New("No TRNG found on platform")
+	}
+
+	if err = ioutil.WriteFile(HwRandomCurrentFile, []byte(selectedRNG), 0644); err != nil {
+		return false, err
+	}
+
+	currentFileData, err := ioutil.ReadFile(HwRandomCurrentFile)
+	if err != nil {
+		return false, err
+	}
+	currentRNG = string(currentFileData)
+
+	if currentRNG != selectedRNG {
+		return false, errors.New("Couldn't select TRNG: " + currentRNG)
+	}
+
+	return true, nil
+}
+
+// UpdateLinuxRandomness seeds random data from
+// /dev/hwrng into /dev/random based on a timer and
+// the entropy pool size
+// Usage:
+// go UpdateLinuxRandomness()
+func UpdateLinuxRandomness(recoverer recovery.Recoverer) error {
+	if good, err := setAvailableTRNG(); err != nil {
+		return err
+	} else if !good {
+		return errors.New("Could not find a good TRNG")
+	}
+
+	hwRng, err := os.OpenFile(HwRandomDevice, os.O_RDONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	rng, err := os.OpenFile(RandomDevice, os.O_APPEND|os.O_WRONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer hwRng.Close()
+		defer rng.Close()
+
+		for {
+			time.Sleep(EntropyFeedTime * time.Second)
```

Comment:
remove time.Second from here. After the change above, EntropyFeedTime is already a valid time.Duration

Reply:
Done